packages/core/src/browser/preferences/preference-service.spec.ts

```diff
@@ -0,0 +1,459 @@
+        const workspaceProvider = getProvider(PreferenceScope.Workspace);
+        const folderProvider = getProvider(PreferenceScope.Folder);
+        userProvider.setPreference('test.number', 1);
+        expect(prefService.get('test.number')).equals(1);
+        workspaceProvider.setPreference('test.number', 0);
+        expect(prefService.get('test.number')).equals(0);
+        folderProvider.setPreference('test.number', 2);
+        expect(prefService.get('test.number')).equals(2);
+
+        // remove property on lower scope
+        folderProvider.setPreference('test.number', undefined);
+        expect(prefService.get('test.number')).equals(0);
+    });
+
+    it('should throw a TypeError if the preference (reference object) is modified', () => {
+        prefSchema.setSchema({
+            properties: {
+                'test.immutable': {
+                    type: 'array',
+                    items: {
+                        type: 'string'
+                    },
+                    scope: 'resource'
+                }
+            }
+        });
+        const userProvider = getProvider(PreferenceScope.User);
+        userProvider.setPreference('test.immutable', [
+            'test', 'test', 'test'
+        ]);
+        const immutablePref: string[] | undefined = prefService.get('test.immutable');
+        expect(immutablePref).to.not.be.undefined;
+        if (immutablePref !== undefined) {
+            expect(() => immutablePref.push('fails')).to.throw(TypeError);
+        }
+    });
+
+    it('should still report the more specific preference even though the less specific one changed', () => {
+        prefSchema.setSchema({
+            properties: {
+                'test.number': {
+                    type: 'number',
+                    scope: 'resource'
+                }
+            }
+        });
+        const userProvider = getProvider(PreferenceScope.User);
+        const workspaceProvider = getProvider(PreferenceScope.Workspace);
+        userProvider.setPreference('test.number', 1);
+        workspaceProvider.setPreference('test.number', 0);
+        expect(prefService.get('test.number')).equals(0);
+
+        userProvider.setPreference('test.number', 4);
+        expect(prefService.get('test.number')).equals(0);
+    });
+
+    describe('overridden preferences', () => {
+
+        it('getPreferences', () => {
+            const { preferences, schema } = prepareServices();
+
+            preferences.set('[json].editor.tabSize', 2, PreferenceScope.User);
+
+            expect(preferences.getPreferences().editor.tabSize).to.equal(4);
+            expect(preferences.getPreferences(undefined, 'json').editor.tabSize).to.equal(4);
+
+            schema.registerOverrideIdentifier('json');
+
+            expect(preferences.getPreferences().editor.tabSize).to.equal(4);
+            expect(preferences.getPreferences(undefined, 'json').editor.tabSize).to.equal(2);
+        });
```

Comment:
It sends duplicate data to language servers:
```ts
        it('toJSON', () => {
            const { preferences } = prepareServices({
                schema: {
                    properties: {
                        'foo.baz': {
                            type: 'number',
                            default: 4
                        },
                        'foor.bar.x': {
                            type: 'boolean',
                            default: true
                        },
                        'foor.bar.y': {
                            type: 'boolean',
                            default: false
                        },
                        'a': {
                            type: 'string',
                            default: 'a'
                        }
                    }
                }
            });

            assert.deepStrictEqual(JSON.stringify(preferences.getPreferences(), undefined, 2), JSON.stringify({
                foo: {
                    baz: 4,
                    bar: {
                        x: true,
                        y: false
                    }
                },
                a: 'a'
            }, undefined, 2), 'there should not be foo.bar.x to avoid sending excessive data to remote clients');
        });
```

Reply:
added that testcase